### PR TITLE
Add IdealStateCalculator.

### PR DESCRIPTION
**Pull request by toregge**
@vekterli : please review

### CODE REVIEW

File: storage/src/vespa/storage/distributor/ideal_state_calculator.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+namespace storage::distributor {
+
+class DistributorInterface;
+class DistributorBucketSpaceRepo;
+
+/*
+ * Class for ideal state calculations with simple hash maps to reuse old
+ * results when they are still valid.
+ */
+class IdealStateCalculator
```
Comment (geirst) [inline storage/src/vespa/storage/distributor/ideal_state_calculator.h:20]: I think we should write separate unit tests for this class.

**Review by vekterli — COMMENTED** [review]
See comment regarding ideal state caching when number of bucket bits exceeds colocation threshold, otherwise LGTM. Agree with @geirst that separate unit tests for the class would be very nice.

File: storage/src/vespa/storage/distributor/distributorcomponent.cpp [hunk 2]
```diff
@@ -256,26 +226,27 @@ DistributorComponent::updateBucketDatabase(
...
     // Ensure that we're not trying to bring any zombie copies into the
     // bucket database (i.e. copies on nodes that are actually unavailable).
-    std::vector<uint16_t> unavailableNodes;
-    enumerateUnavailableNodes(unavailableNodes, bucketSpace.getClusterState(), bucket, changedNodes);
-    if (auto* pending_state = _distributor.pendingClusterStateOrNull(bucket.getBucketSpace())) {
-        enumerateUnavailableNodes(unavailableNodes, *pending_state, bucket, changedNodes);
+    const auto& available_nodes = _ideal_state_calculator.get_available_nodes(bucket.getBucketSpace());
+    bool found_down_node = false;
+    for (const auto& copy : changedNodes) {
+        if (copy.getNode() >= available_nodes.size() || !available_nodes[copy.getNode()]) {
```
[the remark this answers is not in the document]
Comment (toregge) [reply storage/src/vespa/storage/distributor/distributorcomponent.cpp:232]: That will add a hash table lookup for each copy.

**Comment by toregge** [comment]
Closing ticket for code refactoring.  Will move most of IdealStateCalculator to DistributorBucketSpace, to reduce number of needed
hash lookups.

**Pull request closed without merging**